crates/kornia-linalg/src/linalg.rs

```diff
@@ -0,0 +1,509 @@
+    _S.m_20 = S.m_10;
+    _S.m_21 = S.m_20;
+    _S.m_22 = S.m_00;
+
+    S.m_00 = _S.m_00;
+    S.m_10 = _S.m_10;
+    S.m_11 = _S.m_11;
+    S.m_20 = _S.m_20;
+    S.m_21 = _S.m_21;
+    S.m_22 = _S.m_22;
+}
+
+/// Function used to contain the givens permutations and the loop of the jacobi steps controlled by JACOBI_STEPS
+/// Returns the quaternion q containing the cumultative result used to reconstruct S
+pub fn jacobi_eigenanalysis(mut S: Symmetric3x3) -> Mat3 {
+    let mut q = IndexedQuat::new(Quat::from_xyzw(0.0, 0.0, 0.0, 1.0));
+    for _i in 0..JACOBI_STEPS {
+        jacobi_conjugation(0, 1, 2, &mut S, &mut q);
+        jacobi_conjugation(1, 2, 0, &mut S, &mut q);
+        jacobi_conjugation(2, 0, 1, &mut S, &mut q);
+    }
+    return quaternion_to_matrix(&q);
+}
+
+/// Implementation of Algorithm 3
+pub fn sort_singular_values(B: &mut Mat3, V: &mut Mat3) {
+    let mut rho1 = dist2(B.x_axis.x, B.y_axis.x, B.z_axis.x);
+    let mut rho2 = dist2(B.x_axis.y, B.y_axis.y, B.z_axis.y);
+    let mut rho3 = dist2(B.x_axis.z, B.y_axis.z, B.z_axis.z);
+
+    let mut c = rho1 < rho2;
+    cond_neg_swap(c, &mut B.x_axis.x, &mut B.x_axis.y);
+    cond_neg_swap(c, &mut V.x_axis.x, &mut V.x_axis.y);
+    cond_neg_swap(c, &mut B.y_axis.x, &mut B.y_axis.y);
+    cond_neg_swap(c, &mut V.y_axis.x, &mut V.y_axis.y);
+    cond_neg_swap(c, &mut B.z_axis.x, &mut B.z_axis.y);
+    cond_neg_swap(c, &mut V.z_axis.x, &mut V.z_axis.y);
+    cond_swap(c, &mut rho1, &mut rho2);
+
+    c = rho1 < rho3;
+    cond_neg_swap(c, &mut B.x_axis.x, &mut B.x_axis.z);
+    cond_neg_swap(c, &mut V.x_axis.x, &mut V.x_axis.z);
+    cond_neg_swap(c, &mut B.y_axis.x, &mut B.y_axis.z);
+    cond_neg_swap(c, &mut V.y_axis.x, &mut V.y_axis.z);
+    cond_neg_swap(c, &mut B.z_axis.x, &mut B.z_axis.z);
+    cond_neg_swap(c, &mut V.z_axis.x, &mut V.z_axis.z);
+    cond_swap(c, &mut rho1, &mut rho3);
+
+    c = rho2 < rho3;
+    cond_neg_swap(c, &mut B.x_axis.y, &mut B.x_axis.z);
+    cond_neg_swap(c, &mut V.x_axis.y, &mut V.x_axis.z);
+    cond_neg_swap(c, &mut B.y_axis.y, &mut B.y_axis.z);
+    cond_neg_swap(c, &mut V.y_axis.y, &mut V.y_axis.z);
+    cond_neg_swap(c, &mut B.z_axis.y, &mut B.z_axis.z);
+    cond_neg_swap(c, &mut V.z_axis.y, &mut V.z_axis.z);
+}
+
+/// Implementation of Algorithm 4
+pub fn qr_givens_quaternion(a1: f32, a2: f32) -> Givens {
+    // a1 = pivot point on diagonal
+    // a2 = lower triangular entry we want to annihilate
+    let epsilon = SVD_EPSILON; // Assuming _SVD_EPSILON is defined elsewhere
+    let rho = accurate_sqrt(a1 * a1 + a2 * a2);
+
+    let mut g = Givens {
+        ch: (a1.abs() + (f32::max(rho, epsilon))),
+        sh: if rho > epsilon { a2 } else { 0.0 },
+    };
+
+    let b = a1 < 0.0;
+    cond_swap(b, &mut g.sh, &mut g.ch);
+
+    let w = rsqrt(g.ch * g.ch + g.sh * g.sh);
+    g.ch *= w;
+    g.sh *= w;
+
+    return g;
+}
+
+/// Implements a QR decomposition of a Matrix
+pub fn qr_decomposition(B: &mut Mat3) -> QR {
```

Comment:
```suggestion
fn qr_decomposition(B: &mut Mat3) -> QR {
```
make all the "utils" as private functions